### PR TITLE
Et upgrade apr 12th

### PR DESCRIPTION
**Pull request by kanxu526**
1) Allow heterogeneous varaince;
2) Add option for using constraints on threshold or not;

### CODE REVIEW

File: R/ebayesthresh.R [hunk 1]
```diff
@@ -1,6 +1,6 @@
 "ebayesthresh" <-
 function(x, prior = "laplace", a = 0.5, bayesfac = FALSE, sdev = NA, verbose = FALSE, 
-	threshrule = "median")
+	threshrule = "median", threshrestriction = TRUE)
```
Comment (stephens999) [inline R/ebayesthresh.R]: have you updated the help page to explain this parameter?

File: R/wandafromx.R [hunk 2]
```diff
@@ -1,13 +1,14 @@
...
 {
 #  finds the marginal max lik estimators of w and a, using a bivariate optimization
 #
 #  The threshold is constrained to lie between 0 and sqrt ( 2 log (n)) * s
 #
 #  If running R, the routine optim is used; in S-PLUS the routine is nlminb
 #
-	thi <- sqrt(2 * log(length(x))) * s
+	if(threshrestriction) { thi <- sqrt(2 * log(length(x))) * s
+	} else{ thi <- Inf * s }
```
Comment (stephens999) [inline R/wandafromx.R]: dont' really need s here, right? Inf is enough? Or is there a reason?

File: R/laplace.threshzero.R [hunk 3]
```diff
@@ -7,8 +7,6 @@ function(x, s, w, a = 0.5)
 #  only allow a < 20 for input value.
 	a <- min(a, 20)
   xma <- x/s - s*a
-  xma[abs(xma) > 35] <- sign(xma[abs(xma) > 35]) * 35
-  
-	z <- pnorm(xma) - (dnorm(xma) * (1/w + beta.laplace(x, s, a)))/a
+	z <- pnorm(xma) - (dnorm(xma) * (1/w + beta.laplace(x, s, a)))/a/s
```
Comment (stephens999) [inline R/laplace.threshzero.R]: can you clarify this expression by adding parentheses? specifically, is the denominator (a/s) ?
a/b/c is pretty confusing to look at!

File: R/wfromt.R [hunk 4]
```diff
@@ -4,8 +4,13 @@ function(tt, s, prior = "laplace", a = 0.5)
...
 #
 	pr <- substring(prior, 1, 1)
-	if(pr == "l")
-		wi <- (a * pnorm(tt/s - s*a))/dnorm(tt/s - s*a) - beta.laplace(tt, s, a)
+	if(pr == "l"){
+	  tma <- tt/s - s*a
+	  wi <- 1/abs(tma)
+	  tma[tma > 35] <- 35
+	  wi[tma > -35] <- pnorm(tma[tma > -35])/dnorm(tma[tma > -35])
+		wi <- a * s * wi - beta.laplace(tt, s, a)
```
Comment (stephens999) [inline R/wfromt.R]: no indent here...

File: R/postmed.laplace.R [hunk 5]
```diff
@@ -13,9 +13,9 @@ function(x, s, w, a = 0.5)
 	sx <- sign(x)
 	x <- abs(x)
 	xma <- x/s - s*a
-	zz <- (dnorm(xma) * (1/w + beta.laplace(x, s, a)))/a
-	zz[xma > 25] <- 0.5
+	zz <- (dnorm(xma) * (1/w + beta.laplace(x, s, a)))/a /s
```
Comment (stephens999) [inline R/postmed.laplace.R]: parentheses here to clarify expression a/b/c

File: R/negloglik.laplace.R [hunk 6]
```diff
@@ -1,16 +1,20 @@
...
 #
 	a <- xpar[2]
-	w <- wfromt(xpar[1], min(ss), a = a)
-	loglik <- sum(log(1 + w * matrix(beta.laplace(xx, ss, a), nrow=1)))
+	# Calculate the range of w given a, using negative monotonicity between w and t
+	wlo <- wfromt(thi, ss, a = a)
+	whi <- wfromt(tlo, ss, a = a)
+	wlo <- max(wlo)
+	whi <- min(whi)
+	loglik <- sum(log(1 + (xpar[1] * (whi - wlo) + wlo) * beta.laplace(xx, ss, a)))
```
Comment (stephens999) [inline R/negloglik.laplace.R:18]: i think i don't understand quite what is going on here. is there an equation or section in the report that relates to this?

File: R/ebayesthresh.R (hunk 1 above)
Comment (kanxu526) [reply R/ebayesthresh.R]: Sorry, not yet, I'll update it soon.

File: R/wandafromx.R (hunk 2 above)
Comment (kanxu526) [reply R/wandafromx.R]: It's a mistake.

File: R/negloglik.laplace.R (hunk 6 above)
Comment (kanxu526) [reply R/negloglik.laplace.R:18]: Please refer to Equation 2.21 in the report. wlo and whi is the lower and upper bound of weight. wlo is actually 1 here so it's not necessary to write it in this way.

File: R/wfromt.R (hunk 4 above)
Comment (kanxu526) [reply R/wfromt.R]: It might be a typo here.

Comment (stephens999) [reply R/wfromt.R]: so will you fix the couple of typos and then we can merge?

**Comment by kanxu526** [comment]
I have updated all the files.

**Pull request merged**